cpp/src/join/hash_join.cu

```diff
@@ -87,62 +89,81 @@ std::size_t compute_join_output_size(table_device_view build_table,
+
+  auto const comparator_helper = [&](auto device_comparator) {
+    pair_equality equality{device_comparator};
+
+    auto iter = cudf::detail::make_counting_transform_iterator(0, pair_func);
+    if (join == join_kind::LEFT_JOIN) {
+      return hash_table.pair_count_outer(
+        iter, iter + probe_table_num_rows, equality, stream.value());
+    } else {
+      return hash_table.pair_count(iter, iter + probe_table_num_rows, equality, stream.value());
+    }
+  };
 
-  std::size_t size;
-  if constexpr (JoinKind == join_kind::LEFT_JOIN) {
-    size = hash_table.pair_count_outer(iter, iter + probe_table_num_rows, equality, stream.value());
+  if (cudf::detail::has_nested_columns(probe_table)) {
+    auto const device_comparator = row_comparator.equal_to<true>(has_nulls, nulls_equal);
+    return comparator_helper(device_comparator);
   } else {
-    size = hash_table.pair_count(iter, iter + probe_table_num_rows, equality, stream.value());
+    auto const device_comparator = row_comparator.equal_to<false>(has_nulls, nulls_equal);
+    return comparator_helper(device_comparator);
   }
-
-  return size;
 }
 
 /**
  * @brief Probes the `hash_table` built from `build_table` for tuples in `probe_table`,
  * and returns the output indices of `build_table` and `probe_table` as a combined table.
  * Behavior is undefined if the provided `output_size` is smaller than the actual output size.
  *
- * @tparam JoinKind The type of join to be performed.
- *
  * @param build_table Table of build side columns to join.
  * @param probe_table Table of probe side columns to join.
  * @param hash_table Hash table built from `build_table`.
+ * @param join The type of join to be performed
  * @param compare_nulls Controls whether null join-key values should match or not.
  * @param output_size Optional value which allows users to specify the exact output size.
  * @param stream CUDA stream used for device memory operations and kernel launches.
  * @param mr Device memory resource used to allocate the returned vectors.
  *
  * @return Join output indices vector pair.
  */
-template <join_kind JoinKind>
 std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
           std::unique_ptr<rmm::device_uvector<size_type>>>
-probe_join_hash_table(cudf::table_device_view build_table,
-                      cudf::table_device_view probe_table,
-                      cudf::detail::multimap_type const& hash_table,
-                      bool has_nulls,
-                      null_equality compare_nulls,
-                      std::optional<std::size_t> output_size,
-                      rmm::cuda_stream_view stream,
-                      rmm::mr::device_memory_resource* mr)
+probe_join_hash_table(
+  cudf::table_view const& build_table,
+  cudf::table_view const& probe_table,
+  std::shared_ptr<cudf::experimental::row::equality::preprocessed_table> preprocessed_build,
+  std::shared_ptr<cudf::experimental::row::equality::preprocessed_table> preprocessed_probe,
```

Comment:
These pointers can also be `const&`.

Reply:
In general, it's not a recommended practice to pass shared pointers by reference.

Reply:
FWIW, I agree. The copy construct of `shared_ptr` just updates the reference count and that should be the desired behavior, and not passing by reference